iree/schemas/metal_executable_def.fbs

```diff
@@ -30,6 +37,15 @@ table MetalExecutableDef {
   // library.
   entry_points:[string];
 
+  // Threadgroup sizes for each entry point.
+  //
+  // We need this because workgroup size is directly baked inside SPIR-V code,
+  // but in Metal it's specified when dispatching workload. So when cross
+  // compiling SPIR-V to MSL, we need to persist the information here so that
+  // later it can be used for dispatching.
+  // TODO(antiagainst): support SPIR-V specialization constant.
+  threadgroup_sizes:[MetalThreadgroupSize];
```

Comment:
Another place to pass this data along would be in the IR around https://github.com/google/iree/blob/c88982b7df2a16a03839f7568f9c8c6f412eeb1b/iree/compiler/Dialect/HAL/IR/HALOps.td#L1232-L1243

with some different bookkeeping between Metal (where those values are load bearing) and Vulkan/other targets (where the values are redundant), right? Is this approach preferred? Seems like a bit more runtime complexity that we could handle in the compiler.

Reply:
The Metal HAL driver also uses The SPIR-V CodeGen path, which by itself decides the workgroup size. So while Metal itself allows the flexibility to set both the workgroup count and workgroup size, it's not the case anymore because we are sharing the SPIR-V CodeGen path (which are built around the assumption that workgroup size is controlled by the CodeGen instead of runtime). So I think just baking it into the final FlatBuffer is more suitable.

Reply:
We control the SPIR-V codegen path and the full lowering process though, so we don't _need_ to restrict ourselves to what it does today for Vulkan. I'm not saying one approach is better than the other, I'm just wondering what the tradeoffs are and if a different approach would make more sense for Metal.

Reply:
Right. I'm certainly not saying Vulkan's split is better. :D Actually it's more restrictive. My point is that at the moment the whole SPIR-V CodeGen pipeline is built around the concept of specifying the workgroup size and count at different places. To allow Metal to specify the workgroup size at launch time, we need to figure out how to change the assumption there (which is across various passes), if we still want to reuse the SPIR-V CodeGen pipeline, or developing a Metal CodeGen path entirely. Both seems to be quite non-trivial effort which I'd think we don't want to take on immediately. I'm a bit dubious about what the flexibly buys us too: even allowing specifying the workgroup size at launch time, it does not mean we are gonna vary it every time and use some values like (31, 34, 62). Mostly I guess it's still something like (64, 1, 1) or (32, 4, 4) or whatever well defined values that match the GPU and the kernel. But I can certainly be missing something here. :) Basically the above motivates me to put it in the FlatBuffer.